### PR TITLE
add region split

### PR DESCRIPTION
**Pull request by morefreeze**
Add region Split function.
Add ServiceType, because connection need divide into some type.
@c4pt0r @qiuyesuifeng PTAL

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -99,6 +103,9 @@ type client struct {
 func serverNameToAddr(server *proto.ServerName) string {
 	return fmt.Sprintf("%s:%d", server.GetHostName(), server.GetPort())
 }
+func addrAndServiceToCache(addr string, srvType ServiceType) string {
```
Comment (qiuyesuifeng) [inline client.go]: Add an empty space.
addrAndServiceToCache seems not a good name, how about connCacheKey?

File: admin.go [hunk 2]
```diff
@@ -273,3 +274,61 @@ func (c *client) TableExists(tbl string) (bool, error) {
 
 	return found, nil
 }
+
+// Split split region
```
Comment (qiuyesuifeng) [inline admin.go]: split -> splits
Comment end of .

File: admin.go [hunk 3]
```diff
@@ -273,3 +274,61 @@ func (c *client) TableExists(tbl string) (bool, error) {
...
+	tbl = tblOrRegion
+	regions, err := c.GetRegions([]byte(tbl), false)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// split all region
+	for _, region := range regions {
+		err := c.split(region, []byte(splitPoint))
+		if err != nil {
+			return err
```
Comment (qiuyesuifeng) [inline admin.go]: errors.Trace(err)

File: utils.go [hunk 4]
```diff
@@ -10,6 +13,23 @@ func retrySleep(retries int) {
...
 }
 
+func NewRegionSpecifier(regionName string) *proto.RegionSpecifier {
+	return &proto.RegionSpecifier{
+		Type:  proto.RegionSpecifier_REGION_NAME.Enum(),
+		Value: []byte(regionName),
+	}
+}
+
+func FindKey(region *RegionInfo, key []byte) bool {
```
Comment (qiuyesuifeng) [inline utils.go]: Seems FindKey can be un-exported?

File: utils.go [hunk 5]
```diff
@@ -10,6 +13,23 @@ func retrySleep(retries int) {
...
+	if region == nil {
+		return false
+	}
+	// StartKey <= key < EndKey
+	return (region.StartKey == nil || bytes.Compare(region.StartKey, key) <= 0) &&
+		(region.EndKey == nil || len(region.EndKey) == 0 ||
```
Comment (qiuyesuifeng) [inline utils.go]: `region.EndKey == nil` is needless, for if `region.EndKey == nil` then `len(region.EndKey) == 0`.

File: utils.go [hunk 6]
```diff
@@ -1,8 +1,11 @@
 package hbase
 
 import (
+	"bytes"
 	"time"
 
+	"github.com/pingcap/go-hbase/proto"
+
```
Comment (qiuyesuifeng) [inline utils.go]: Remove empty line.

File: scan.go [hunk 7]
```diff
@@ -334,8 +321,8 @@ func (s *Scan) nextBatch() int {
 		log.Errorf("scan next batch failed - [startKey=%q], %v", startKey, errors.ErrorStack(err))
 	}
 
-	// current region get 0 data, switch next region
-	if len(rs) == 0 && s.nextStartKey != nil {
+	// Current region get 0 data, try switch next region.
```
Comment (qiuyesuifeng) [inline scan.go]: switch to

File: del_test.go [hunk 8]
```diff
@@ -72,6 +72,7 @@ func (s *HBaseDelTestSuit) TestDelWithClient(c *C) {
 	c.Assert(ok, IsTrue)
 	c.Assert(err, IsNil)
 
+	// check it
```
Comment (qiuyesuifeng) [inline del_test.go]: Seems this comment line is meaningless.

File: conn.go [hunk 9]
```diff
@@ -98,22 +119,23 @@ func readPayloads(r io.Reader) ([][]byte, error) {
...
 
-func newConnection(addr string, isMaster bool) (*connection, error) {
+func newConnection(addr string, srvType ServiceType) (*connection, error) {
 	conn, err := net.Dial("tcp", addr)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-
+	if _, ok := ServiceString[srvType]; !ok {
+		return nil, errors.New("unexpected service type")
```
Comment (qiuyesuifeng) [inline conn.go]: It is better to add srvType in log.

File: conn.go [hunk 10]
```diff
@@ -15,6 +15,27 @@ import (
 	"github.com/pingcap/go-hbase/proto"
 )
 
+type ServiceType int
+
+const (
+	MasterMonitorService = iota
```
Comment (qiuyesuifeng) [inline conn.go]: iota + 1

File: conn.go [hunk 11]
```diff
@@ -15,6 +15,27 @@ import (
 	"github.com/pingcap/go-hbase/proto"
 )
 
+type ServiceType int
```
Comment (qiuyesuifeng) [inline conn.go]: It is better to use `type ServiceType byte`.

File: client.go [hunk 12]
```diff
@@ -394,6 +414,27 @@ func (c *client) LocateRegion(table, row []byte, useCache bool) (*RegionInfo, er
 	return nil, errors.Errorf("Couldn't find the region: [table=%s] [row=%q] [region_row=%q]", table, row, regionRow)
 }
 
+func (c *client) GetRegions(table []byte, useCache bool) ([]*RegionInfo, error) {
+	var regions []*RegionInfo
+	startKey := []byte("")
+	// get first region
+	region, err := c.LocateRegion(table, []byte(startKey), useCache)
+	if err != nil {
+		return nil, errors.Errorf("Couldn't find any region")
```
Comment (qiuyesuifeng) [inline client.go]: Add more detailed trace info in log.

File: client.go [hunk 13]
```diff
@@ -182,9 +189,9 @@ func (c *client) init() error {
 }
 
 // get connection
-func (c *client) getConn(addr string, isMaster bool) (*connection, error) {
+func (c *client) getConn(addr string, srvType ServiceType) (*connection, error) {
 	c.mu.RLock()
-	conn, ok := c.cachedConns[addr]
+	conn, ok := c.cachedConns[addrAndServiceToCache(addr, srvType)]
```
Comment (qiuyesuifeng) [inline client.go]: Use a variable for addrAndServiceToCache(addr, srvType).

File: client.go [hunk 14]
```diff
@@ -160,13 +167,13 @@ func (c *client) init() error {
 	}
 
 	log.Debug("connect root region server...", c.rootServerName)
-	conn, err := newConnection(serverNameToAddr(c.rootServerName), false)
+	conn, err := newConnection(serverNameToAddr(c.rootServerName), ClientService)
```
Comment (qiuyesuifeng) [inline client.go]: Use a variable for serverNameToAddr(c.rootServerName), ClientService).

File: admin_test.go [hunk 15]
```diff
@@ -91,3 +100,58 @@ func (s *AdminTestSuit) TestCreateTableAsync(c *C) {
...
+	c.Assert(err, IsNil)
+	c.Assert(regions, HasLen, 4)
+}
+
+func (s *AdminTestSuit) TestTableAutoSplit(c *C) {
+	cli, _ := NewClient(getTestZkHosts(), "/hbase")
+	regions, err := cli.GetRegions([]byte(s.tableName), false)
+	c.Assert(err, IsNil)
+	c.Assert(regions, HasLen, 4)
+	for c := 'b'; c < 'f'; c++ {
```
Comment (qiuyesuifeng) [inline admin_test.go]: c is not a good local variable.

File: admin_test.go [hunk 16]
```diff
@@ -91,3 +100,58 @@ func (s *AdminTestSuit) TestCreateTableAsync(c *C) {
...
+		// b_0, b_1, ...
+		// if insert few row, it may not be splited even invoke Split explict
+		for i := 0; i < 10000; i++ {
+			p := NewPut([]byte(fmt.Sprintf("%c_%d", c, i)))
+			p.AddStringValue("cf", "c", fmt.Sprintf("%c%c_%d", c, c, i))
+			cli.Put(s.tableName, p)
```
Comment (qiuyesuifeng) [inline admin_test.go]: Add error check.

File: client.go (hunk 14 above)
Comment (morefreeze) [inline client.go]: serverNameToAddr(c.rootServerName) is a thing
ClientService is anothor thing
they both are arguments of newConnection

File: client.go (hunk 1 above)
Comment (morefreeze) [inline client.go]: because it used to be used in cachedConns, so how about `cachedConnKey`?

Comment (qiuyesuifeng) [inline client.go]: Sounds better.

File: client.go (hunk 14 above)
Comment (morefreeze) [inline client.go]: ok

**Comment by morefreeze** [comment]
@qiuyesuifeng @c4pt0r PTAL

File: client.go [hunk 17]
```diff
@@ -9,8 +9,10 @@ import (
 	"sync"
 	"time"
 
-	pb "github.com/golang/protobuf/proto"
 	"github.com/juju/errors"
+
```
Comment (qiuyesuifeng) [inline client.go]: It is better to stay the same.
The import format looks like 
```
official packages
empty line
other packages
```

**Comment by qiuyesuifeng** [comment]
@c4pt0r PTAL

File: client.go [hunk 18]
```diff
@@ -394,6 +416,27 @@ func (c *client) LocateRegion(table, row []byte, useCache bool) (*RegionInfo, er
...
+	// Get first region.
+	region, err := c.LocateRegion(table, []byte(startKey), useCache)
+	if err != nil {
+		return nil, errors.Errorf("couldn't find any region: [table=%s] [useCache=%t]", table, useCache)
+	}
+	regions = append(regions, region)
+	startKey = region.EndKey
+	for len(startKey) > 0 {
+		region, err = c.LocateRegion(table, []byte(startKey), useCache)
+		if err != nil {
```
Comment (c4pt0r) [inline client.go]: untreated error?

Comment (morefreeze) [inline client.go]: I will add `return nil, errors.Trace(err)`

**Comment by c4pt0r** [comment]
LGTM

**Pull request merged**